File: Conceitos de LPs/Subprogramas/Corrotinas/corrotinas.cpp
//C++
//C++20 ou superior tem suporte nativo para co-rotinas com a palavra-chave co_await, co_yield e co_return.

#include <iostream>
#include <coroutine>
#include <chrono>
#include <thread>

using namespace std;

struct Task {
    struct promise_type {
        Task get_return_object() { return {}; }
        suspend_never initial_suspend() { return {}; }
        suspend_never final_suspend() noexcept { return {}; }
        void return_void() {}
        void unhandled_exception() {}
    };
};

Task tarefa1() {
    for (int i = 1; i <= 3; i++) {
        cout << "Tarefa 1 - execução " << i << endl;
        this_thread::sleep_for(chrono::seconds(1));
        co_await suspend_always{};
    }
}

Task tarefa2() {
    for (int i = 1; i <= 2; i++) {
        cout << "Tarefa 2 - execução " << i << endl;
        this_thread::sleep_for(chrono::seconds(2));
        co_await suspend_always{};
    }
}

int main() {
    auto t1 = tarefa1();
    auto t2 = tarefa2();
}

File: Conceitos de LPs/Subprogramas/Corrotinas/corrotinas.js
//JavaScript

function esperar(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function tarefa1() {
    for (let i = 0; i < 3; i++) {
        console.log(`Tarefa 1 - execução ${i + 1}`);
        await esperar(1000);
    }
}

async function tarefa2() {
    for (let i = 0; i < 2; i++) {
        console.log(`Tarefa 2 - execução ${i + 1}`);
        await esperar(2000);
    }
}

async function main() {
    await Promise.all([tarefa1(), tarefa2()]);
}

main();
